src/Mvc/Mvc.RazorPages/src/Infrastructure/DynamicPageEndpointSelector.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+using Microsoft.AspNetCore.Routing;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    internal class DynamicPageEndpointSelector : IDisposable
+    {
+        private readonly PageActionEndpointDataSource _dataSource;
+        private readonly DataSourceDependentCache<ActionSelectionTable<RouteEndpoint>> _cache;
+
+        public DynamicPageEndpointSelector(PageActionEndpointDataSource dataSource)
+        {
+            if (dataSource == null)
+            {
+                throw new ArgumentNullException(nameof(dataSource));
+            }
+
+            _dataSource = dataSource;
+            _cache = new DataSourceDependentCache<ActionSelectionTable<RouteEndpoint>>(dataSource, Initialize);
+        }
+
+        private ActionSelectionTable<RouteEndpoint> Table => _cache.EnsureInitialized();
+
+        // This is async because the page version will be async. We don't want to put ourselves in a
```

Comment:
> This is async because the page version will be async.

This is the page version. Neither the controller selector nor the page do async work. Should they be synchronous?

Reply:
I'll remove the comment - also this will be async once I use your new thing.

 I **insist** that these be asynchronous because I anticipate that they will be public. I do not want them to be made synchronous and then proliferate usage all over the world.

Reply:
GAH. You are the correct. These are no longer async.